src/Microsoft.OData.Client/Serialization/HttpWebRequestMessage.cs

```diff
@@ -286,179 +355,113 @@ public override IAsyncResult BeginGetRequestStream(AsyncCallback callback, objec
-                    httpWebRequest.Headers[XmlConstants.HttpContentLength] = XmlConvert.ToString(contentLength);
-                }
-                catch (ArgumentException)
-                {
-                    SettingContentLengthOnHeadersCollectionValidOnCurrentPlatform = false;
+                    _requestMessage.Content.Headers.Add(contentHeader.Key, contentHeader.Value);
                 }
             }
-#else
-            httpWebRequest.ContentLength = contentLength;
-#endif
+
+            _requestMessage.Method = new HttpMethod(_effectiveHttpMethod);
+
+            var send = _client.SendAsync(_requestMessage);
+            return send;
         }
 
-        /// <summary>
-        /// Sets the accept charset.
-        /// </summary>
-        /// <param name="httpWebRequest">The HTTP web request.</param>
-        /// <param name="headerValue">The header value.</param>
-        internal static void SetAcceptCharset(HttpWebRequest httpWebRequest, string headerValue)
+        private static IDictionary<string, string> HttpHeadersToStringDictionary(HttpHeaders headers)
         {
-#if !PORTABLELIB
-            httpWebRequest.Headers[XmlConstants.HttpAcceptCharset] = headerValue;
-#endif
-#if PORTABLELIB
-            if (!IsRunningOnSilverlight)
-            {
-                httpWebRequest.Headers[XmlConstants.HttpAcceptCharset] = headerValue;
-            }
-#endif
+            return headers.ToDictionary((h1) => h1.Key, (h2) => string.Join(",", h2.Value));
         }
 
-        /// <summary>
-        /// Attempts to set the UserAgent property if it exists other wise returns false
-        /// </summary>
-        /// <param name="httpWebRequest">The http web request.</param>
-        /// <param name="headerValue">The value of the user agent.</param>
-        internal static void SetUserAgentHeader(HttpWebRequest httpWebRequest, string headerValue)
+        /// <summary> Returns the underlying HttpRequestMessage </summary>
+        public System.Net.Http.HttpRequestMessage HttpRequestMessage
         {
-#if !PORTABLELIB
-            httpWebRequest.UserAgent = headerValue;
-#endif
-#if PORTABLELIB
-            if (IsRunningOnSilverlight || !SettingUserAgentPropertyValidOnCurrentPlatform)
+            get
             {
-                return;
+                return _requestMessage;
             }
+        }
 
-            bool headerSet = false;
+        private static HttpWebResponseMessage ConvertHttpClientResponse(HttpResponseMessage response)
+        {
+            var allHeaders = HttpHeadersToStringDictionary(response.Headers).Concat(HttpHeadersToStringDictionary(response.Content.Headers));
+            return new HttpWebResponseMessage(
+                allHeaders.ToDictionary((h1) => h1.Key, (h2) => h2.Value),
+                (int)response.StatusCode,
+                () =>
+                {
+                    var task = response.Content.ReadAsStreamAsync();
+                    task.Wait();

```

Comment:
`task.Wait()` is thread blocking. As Billy suggested, we should consider moving to `async/await` pattern in the future.

Read the first answer https://stackoverflow.com/questions/17432306/task-factory-startnew-vs-task-factory-fromasync

Reply:
will add a work item for this